File: Igatha/Sensors/SensorType.swift
//
//  SensorType.swift
//  Igatha
//
//  Created by Nizar Mahmoud on 12/10/2024.
//

import Foundation

enum SensorType {
    case accelerometer
    case gyroscope
    case barometer
}

public protocol AnySensor: AnyObject {
    var isAvailable: Bool { get }
    
    func startUpdates()
    func stopUpdates()
}

protocol Sensor: AnySensor {
    var delegate: SensorDelegate? { get }
    
    associatedtype T
    var sensor: T { get }
    
    var threshold: Double { get }
}

protocol SensorDelegate: AnyObject {
    func sensorExceededThreshold(sensorType: SensorType, eventTime: Date)
}

File: android/app/src/main/java/com/nizarmah/igatha/sensor/SensorType.kt
package com.nizarmah.igatha.sensor

import android.hardware.Sensor
import kotlinx.coroutines.flow.SharedFlow

enum class SensorType {
    ACCELEROMETER,
    GYROSCOPE,
    BAROMETER
}

interface AnySensor {
    val isAvailable: Boolean

    fun startUpdates()
    fun stopUpdates()
}

interface Sensor : AnySensor {
    val sensor: Sensor?
    val threshold: Double
    val sensorType: SensorType

    // A shared flow to emit events when the threshold is exceeded
    val events: SharedFlow<SensorCapturedEvent>
}

data class SensorCapturedEvent(
    val sensorType: SensorType,
    val eventTime: Long
)
